### ISSUE TITLE
How to have completion for Custom Resource Definitions

### ISSUE THREAD

**Issue by apupier** [Azure/vscode-kubernetes-tools#917]
based on https://github.com/Azure/vscode-kubernetes-tools/pull/583#issue-298696614 i would expect to have completion in this case:
```
apiVersion: camel.apache.org/v1alpha1
kind: KameletBinding
metadata:
  name: bitcoin-to-kafka
spec:
  
```
The kameletBinding CRD deployed 
![Screenshot from 2021-03-30 15-14-19](https://user-images.githubusercontent.com/1105127/113000955-9c534680-9170-11eb-9a47-ab8fc89b501f.png)

I would expect the "integration" completion https://github.com/apache/camel-k/blob/9e384110be106b4970af2e967e28e0877fa78359/config/crd/bases/camel.apache.org_kameletbindings.yaml#L66 but it is not the case.

I also tried to associate the "kubernetes" schema to the file, but in this case I have this error on the "kind" value:
```
Value is not accepted. Valid values: "StorageVersion", "DaemonSet", "Deployment", "ReplicaSet", "StatefulSet", "TokenRequest", "TokenReview", "LocalSubjectAccessReview", "SelfSubjectAccessReview", "SelfSubjectRulesReview", "SubjectAccessReview", "HorizontalPodAutoscaler", "Scale", "Job", "CronJob", "CertificateSigningRequest", "Lease", "LimitRange", "Namespace", "Node", "PersistentVolume", "PersistentVolumeClaim", "Pod", "ReplicationController", "ResourceQuota", "Service", "Ingress", "FlowSchema", "PriorityLevelConfiguration", "IngressClass", "NetworkPolicy", "RuntimeClass", "PodDisruptionBudget", "PodSecurityPolicy", "CSIDriver", "CSINode", "VolumeAttachment", "CustomResourceDefinition", "APIService".yaml-schema: https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/v1.20.5-standalone-strict/_definitions.json(1)
```

What am i missing to have the completion/validation/hover available?

**Comment by lstocchi** [comment]
Quick answer: I tried it and it seems working for me (you can see the json schema file used at the top in the image) so my guess is that you already had the file opened while vscode was fetching the schemas.
![image](https://user-images.githubusercontent.com/49404737/113030392-d4b54d80-918d-11eb-9586-9128e6b972e6.png)

The problem is that this doesn't solve your issue. The schema fetched is 
```
{
  "id": "camel.apache.org/v1alpha1@KameletBinding",
  "apiVersion": "camel.apache.org/v1alpha1",
  "kind": "KameletBinding",
  "type": "object",
  "name": "org.apache.camel.v1alpha1.KameletBinding"
}
```
which is quite useless. 
It looks to me that we only fetch the swagger api from the cluster and maybe this is the result we get. I need to dig a bit more and see how this can be improved. I'll let you know as soon as i have news. Not sure if @itowlson has already an answer for this.

**Comment by itowlson** [comment]
@apupier What cluster version are you on?  From my digging, k8s should be providing us full CRD structural schemas in the Swagger from 1.16 onwards.  There is a slightly enigmatic restriction to "Structural Schemas" though - I wonder if that CRD schema doesn't count as 'structural.'

https://v1-17.docs.kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2

https://v1-18.docs.kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-graduated-or-deprecated-features

**Comment by apupier** [comment]
@itowlson I'm using `minikube version: v1.18.1` with the default Kubernetes version which is `Kubernetes v1.20.2`

**Comment by apupier** [comment]
> I wonder if that CRD schema doesn't count as 'structural.'

From this [blogpost](https://kubernetes.io/blog/2019/06/20/crd-structural-schema/#conclusion):

> Structural schemas are the future of CRDs. apiextensions.k8s.io/v1 will require them.

The apiversion used is the one mentioned

https://github.com/apache/camel-k/blob/9e384110be106b4970af2e967e28e0877fa78359/config/crd/bases/camel.apache.org_kameletbindings.yaml#L18

so I guess that the schema is "structural".

**Comment by apupier** [comment]
> The apiversion used is the one mentioned
> 
> https://github.com/apache/camel-k/blob/9e384110be106b4970af2e967e28e0877fa78359/config/crd/bases/camel.apache.org_kameletbindings.yaml#L18
> 
> so I guess that the schema is "structural".

I also mixed a bunch of things :-)
Finally made it working, several things:
- it seems that a structural schema is required. For my initial example of Camel K, it wil be available with Camel K 1.4.0 thanks to https://github.com/apache/camel-k/pull/1914 . I initially tried with Camel K 1.3.2.
- I tried to close/reopen the editor without success a first time. I reloaded VS Code and then this time the compeltion was available. Matter of timing? or really need to reload VS Code? i'm wondering how to improve that. Maybe a codelens to tell user that computation is in progress? or that a new schema is available but requires a reopening of the file?
- if i understand well, the PR provided by Luca allows to have schema even for older CRDs, right?

**Comment by itowlson** [comment]
Re needing to reload: there is indeed a timing issue due to one of our underlying components that we haven't found a workaround for yet.  I *think* it may be enough to close and reopen the document, but my memory is a bit fogged on this!  In any case, as you say, Luca's approach may be more broadly applicable and I'll try to get his PR reviewed today.
